javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,153 @@
+    static int firstDigit(int num, int depthShift) { return num >> depthShift; }
+    static int digit(int num, int depthShift)      { return lastDigit(firstDigit(num, depthShift)); }
+    static int lastDigit(int num)                  { return num & (-1 >>> -BRANCHING_BASE); }
+
+    private static final long serialVersionUID = 1L;
+
+    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(Arrays.empty(), 0, 0, 0);
+    @SuppressWarnings("unchecked")
+    static <T> BitMappedTrie<T> empty() { return (BitMappedTrie<T>) EMPTY; }
+
+    private final Object[] array;
+    private final int offset, length;
+    private final int depthShift;
+
+    BitMappedTrie(Object[] array, int offset, int length, int depthShift) {
+        this.array = array;
+        this.offset = offset;
+        this.length = length;
+        this.depthShift = depthShift;
+
+        assert length() <= treeSize(branchingFactor(), depthShift);
+    }
+
+    private static int treeSize(int branchCount, int depthShift) {
+        final int fullBranchSize = 1 << depthShift;
+        return branchCount * fullBranchSize;
+    }
+
+    BitMappedTrie<T> prependLeaf(T leading) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(leading), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift, offset = this.offset;
+            if (isFullLeft()) {
+                array = copyUpdate(Arrays.empty(), branchingFactor() - 1, array);
+                depthShift += BRANCHING_BASE;
+                offset = treeSize(branchingFactor() - 1, depthShift);
+            }
+
+            offset -= 1;
+            array = modifyLeaf(array, depthShift, offset, NODE_MODIFIER, LEAF_MODIFIER.apply(leading));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullLeft() { return offset == 0; }
+
+    BitMappedTrie<T> appendLeaf(T trailing) {
+        final int newSize = length() + 1;
+        if (length() == 0) {
+            return new BitMappedTrie<>(asArray(trailing), offset, newSize, depthShift);
+        } else {
+            Object[] array = this.array;
+            int depthShift = this.depthShift;
+            if (isFullRight(newSize)) {
+                array = asArray(array);
+                depthShift += BRANCHING_BASE;
+            }
+
+            array = modifyLeaf(array, depthShift, offset + length(), NODE_MODIFIER, LEAF_MODIFIER.apply(trailing));
+            return new BitMappedTrie<>(array, offset, newSize, depthShift);
+        }
+    }
+    private boolean isFullRight(int newSize) { return (offset + newSize) > treeSize(branchingFactor(), depthShift); }
+
+    private static final NodeModifier NODE_MODIFIER = (o, i) -> copy(o, i + 1);
+    private static final Function<Object, NodeModifier> LEAF_MODIFIER = element -> ((o, i) -> copyUpdate(o, i, element));
```

Comment:
You could replace this with a simple method:
```
    private NodeModifier leafModifier(T element) {
        return (o, i) -> copyUpdate(o, i, element);
    }
```

Invocations of the function `LEAF_MODIFIER.apply(foo)` could be transformed to simple method calls: `leafModifier(foo)`.

Since you don't actually store any of the `NodeModifier` instances, there is no caching in place (as opposed to `NODE_MODIFIER`), so I don't see any such performance disadvantages for what I'm proposing.